### DISCUSSION TITLE
Managing typescript & react versions in a monorepo

### DISCUSSION THREAD

**Discussion by ron23** [pnpm/pnpm#6174] [Q&A] [answered]
I'm trying to migrate over a repo with multiple apps into a monorepo using workspaces and had some questions around TS & React and trying to learn the best practices.

1. Can I have different apps have different typescript version? I assume not because the IDE doesn't play with multiple versions, so what's the best practice here?
2. Should I have typescript as dev dependency at the root? I see the pnpm repo uses that approach, but I can seem to run `tsc` on a package. I also get peerDepndency warnings when I do that.
3. React and @types/react. If I have a shared package that uses react 16. Can my consuming app use React 18? Running into a bunch of issues around that. There are also transient dependencies on @types/react which causes mismatches. What's the best practice here?

Thanks!

**Comment by gluxon** [comment, marked as answer]
> Can I have different apps have different typescript version? I assume not because the IDE doesn't play with multiple versions, so what's the best practice here?

In theory it is possible for each app to use a different TypeScript version, but this would make the monorepo very difficult to work in due to the IDE constraint you've noticed. I think the easiest migration path would be to upgrade each application's repo to a common version of TypeScript before merging them into a singular monorepo.

> Should I have typescript as dev dependency at the root? I see the pnpm repo uses that approach, but I can seem to run tsc on a package. I also get peerDepndency warnings when I do that.

How does `tsc` get ran? Setting up a `compile` `package.json` script and simply setting that to `tsc` there should work. I'd also imagine `pnpm exec tsc` to work anywhere in the repo if the `typescript` dependency is added to the root `package.json`.

> React and @types/react. If I have a shared package that uses react 16. Can my consuming app use React 18? Running into a bunch of issues around that. There are also transient dependencies on @types/react which causes mismatches. What's the best practice here?

Similar to TypeScript, while it's possible to have different packages on different versions of React, managing multiple versions is very difficult. I do think the best practice for monorepos is to have as many consistent versions of each dependency as possible. That does mean some up front migration work is necessary, but I think it'll save a lot of time for you and your team in the future.

For React specifically, the consuming app will need to render the shared package on React 16 in an independent React root since React expects to be at 1 consistent version in a project. This would be one of the many painful parts of separate React instances on the same web page.

**Reply by ron23** [reply]
Thank you for the detailed answer.
I guess one version constraint is a double edge sword.
I'm working on a large codebase, with multiple teams & projects and having flexibility around versions is quite beneficial.
eg we don't want to touch deprecated projects (KTLO mode), or have to coordinate a migration continuously when one app wants to bump it's version.

None of it is pnpm problem but wanted to call it out. Thanks again!
